splash/browser_tab.py

```diff
@@ -731,3 +730,64 @@ def log(self, message, min_level=None):
 
         message = "[%s] %s" % (self.uid, message)
         log.msg(message, system='render')
+
+
+def _render_web_page_to_image(web_page, logger):
+    """
+    Render a QWebPage into a QImage.
+
+    This function works around bugs in QPaintEngine that surface when the
+    device is larger than 32k pixels in either dimension.
+
+    """
+    size = web_page.viewportSize()
+
+    # A bug is triggered (on my box) when width or height is >= 32768.
+    tile_maxsize = 16384
+    if size.width() < tile_maxsize and size.height() < tile_maxsize:
+        image = QImage(size, QImage.Format_ARGB32)
+        painter = QPainter(image)
+        web_page.mainFrame().render(painter)
+        return image
+
+    logger.log("png renderer: viewport too large (%s), using tiled rendering" %
+               size, min_level=2)
+    # One bug is worked around by rendering the page one tile at a time onto a
+    # small-ish temporary image.  The magic happens in viewport-window
+    # transformation:
+    #
+    # - Sizes of tile painter viewport and tile painter window match
+    #   webpage viewport size to avoid rescaling.
+    # - Tile painter window is moved appropriately so that tile region is
+    #   overlayed onto the temporary image.
+    tile_hsize = min(tile_maxsize, size.width())
+    tile_vsize = min(tile_maxsize, size.height())
+    htiles = 1 + (size.width() - 1) // tile_hsize
+    vtiles = 1 + (size.height() - 1) // tile_vsize
+    tile_image = QImage(QSize(tile_hsize, tile_vsize), QImage.Format_ARGB32)
+    tile_painter = QPainter(tile_image)
+    viewport = QRect(0, 0, size.width(), size.height())
+    tile_painter.setViewport(viewport)
+
+    # The other bug manifests itself when you do painter.drawImage trying to
+    # concatenate tiles onto a single image and once you reach 32'768 along
+    # either dimension all of a sudden drawImage simply stops drawing anything.
+    # The simplest workaround that comes to mind is to use pillow for pasting
+    # images.
+    pil_image = Image.new(mode='RGBA',
+                          size=(size.width(), size.height()))
+    for i in xrange(htiles):
+        for j in xrange(vtiles):
+            left, top = i * tile_hsize, j * tile_vsize
+            tile_painter.setWindow(viewport.translated(left, top))
+            logger.log("Rendering with window=%s" % tile_painter.window(),
+                       min_level=2)
+            web_page.mainFrame().render(tile_painter)
+            pil_tile_image = qimage_to_pil_image(tile_image)
+
+            logger.log("Pasting rendered tile to coords: %s" %
+                       ((left, top),),
+                       min_level=2)
+            pil_image.paste(pil_tile_image, (left, top))
+    tile_painter.end()
+    return pil_image_to_qimage(pil_image)
```

Comment:
.. and it seems it is QImage creation which uses 4GB of memory; building of PIL image tops at 360MB if a smaller tile size is used (1024px). For 512px tiles top memory usage is 280MB. I suggest to use smaller tile size like this (or maybe larger, to fit common screen widths) because it seems Qt is a memory hog here - with 16384 tile it uses more than 2GB when rendering. With 512px tiles rendering is about 2x slower than with 16K tiles; we need to find a balance.